src/index/base.cpp

```diff
@@ -262,13 +264,27 @@ void BaseIndex::BlockConnected(const std::shared_ptr<const CBlock>& block, const
                       best_block_index->GetBlockHash().ToString());
             return;
         }
-        if (best_block_index != pindex->pprev && !Rewind(best_block_index, pindex->pprev)) {
+
+        // If best_block_index is an ancestor of pindex, then this is a block that we've already
+        // processed. This can occur if the sync thread is almost done catching up, the tip
+        // updates to the next block, and the sync thread finishes catching up to tip before
+        // the BlockConnected event is received. If we've already seen it, there's no reason to
+        // call Rewind, WriteBlock, and SetBestBlockIndex.
+        if (best_block_index->GetAncestor(pindex->nHeight) == pindex) {
+            skip_write = true;
```

Comment:
In commit "index: ignore BlockConnected if pindex is in chain" (270579f85fbe8cf97a9f506a258ad3ed865fcbeb)

Seems like code would be more straightforward without `skip_write` variable. Maybe just:

```c++
if (best_block_index->GetAncestor(pindex->nHeight) == pindex) {
    LogPrintf("%s: WARNING: Block %s does is already-processed ancestor of " /* Continued */
              "known best chain (tip=%s); not updating index\n",
              __func__, pindex->GetBlockHash().ToString(),
              best_block_index->GetBlockHash().ToString());
    return;
}
```